overlord/snapstate/snapstate.go

```diff
@@ -1246,23 +1346,63 @@ func infoForType(st *state.State, snapType snap.Type) (*snap.Info, error) {
 		if typ != snapType {
 			continue
 		}
-		return snapst.CurrentInfo()
+		si, err := snapst.CurrentInfo()
+		if err != nil {
+			return nil, err
+		}
+		res = append(res, si)
 	}
 
-	return nil, state.ErrNoState
+	if len(res) == 0 {
+		return nil, state.ErrNoState
+	}
+
+	return res, nil
+}
+
+func infoForType(st *state.State, snapType snap.Type) (*snap.Info, error) {
+	res, err := infoForTypes(st, snapType)
+	if err != nil {
+		return nil, err
+	}
+	return res[0], nil
 }
 
 // GadgetInfo finds the current gadget snap's info.
 func GadgetInfo(st *state.State) (*snap.Info, error) {
 	return infoForType(st, snap.TypeGadget)
 }
 
-// CoreInfo finds the current OS snap's info.
-func CoreInfo(st *state.State) (*snap.Info, error) {
-	return infoForType(st, snap.TypeOS)
-}
-
 // KernelInfo finds the current kernel snap's info.
 func KernelInfo(st *state.State) (*snap.Info, error) {
 	return infoForType(st, snap.TypeKernel)
 }
+
+// CoreInfo finds the current OS snap's info. If both
+// "core" and "ubuntu-core" is installed then "core"
+// is preferred. Different core names are not supported
+// currently and will result in an error.
+func CoreInfo(st *state.State) (*snap.Info, error) {
+	res, err := infoForTypes(st, snap.TypeOS)
+	if err != nil {
+		return nil, err
+	}
+
+	var find = func(needle string) *snap.Info {
+		for _, si := range res {
+			if si.Name() == needle {
+				return si
+			}
+		}
+		return nil
+	}
+
+	if si := find("core"); si != nil {
+		return si, nil
+	}
+	if si := find("ubuntu-core"); si != nil {
+		return si, nil
+	}
```

Comment:
I can adjust the code accordingly if you prefer this style of loop. I would prefer if we return a different error than state.ErrNoState if we can not find "core" or "ubuntu-core". Otherwise we can not distinguish the case where we have no core snap installed and the case when we have one with an unexpected name.